File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter11/ConditionalClockButton/ConditionalClockButton/TwelveHourClock.cpp
#include "pch.h"
#include "TwelveHourClock.h"

using namespace ConditionalClockButton;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Globalization;
using namespace Windows::Globalization::DateTimeFormatting;
using namespace Windows::UI;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media;

TwelveHourClock::TwelveHourClock()
{
    // Initialize for Hour value of 0
    int hour12 = 12;
    bool isAm = true;
    bool isPm = false;
}

void TwelveHourClock::Hour12::set(int value)
{
    if (hour12 != value)
    {
        hour12 = value;
        OnPropertyChanged("Hour12");
    }
}

void TwelveHourClock::IsAm::set(bool value)
{
    if (isAm != value)
    {
        isAm = value;
        OnPropertyChanged("IsAm");
    }
}

void TwelveHourClock::IsPm::set(bool value)
{
    if (isPm != value)
    {
        isPm = value;
        OnPropertyChanged("IsPm");
    }
}

void TwelveHourClock::OnPropertyChanged(String^ propertyName)
{
    if (propertyName == "Hour")
    {
        this->Hour12 = (this->Hour - 1) % 12 + 1;
        this->IsAm = this->Hour < 12;
        this->IsPm = !this->IsAm;
    }

    ClockBase::OnPropertyChanged(propertyName);
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter11/ConditionalClockButton/ConditionalClockButton/TwelveHourClock.cs
namespace ConditionalClockButton
{
    public class TwelveHourClock : ClockButton.Clock
    {
        // Initialize for Hour value of 0
        int hour12 = 12;
        bool isAm = true;
        bool isPm = false;

        public int Hour12
        {
            set { SetProperty<int>(ref hour12, value); }
            get { return hour12; }
        }

        public bool IsAm
        {
            set { SetProperty<bool>(ref isAm, value); }
            get { return isAm; }
        }

        public bool IsPm
        {
            set { SetProperty<bool>(ref isPm, value); }
            get { return isPm; }
        }

        protected override void OnPropertyChanged(string propertyName)
        {
            if (propertyName == "Hour")
            {
                this.Hour12 = (this.Hour - 1) % 12 + 1;
                this.IsAm = this.Hour < 12;
                this.IsPm = !this.IsAm;
            }

            base.OnPropertyChanged(propertyName);
        }
    }
}
